### ISSUE TITLE
Fl::focus(Fl_Widget*) should be independent of Fl_Widget::visible_focus()

### ISSUE THREAD

**Issue by Albrecht-S** [fltk/fltk#366]
Setting the focus to a widget that has disabled keyboard navigation by calling `Fl_Widget::clear_visible_focus()` or `Fl_Widget::visible_focus(0)` fails because `Fl::focus(Fl_Widget*)` tests the `Fl_Widget::visible_focus()` attribute. This should not be done since `Fl::focus(Fl_Widget*)` should set the focus to the widget unconditionally.

[Documentation of `Fl::focus(Fl_Widget*)`](https://fltk.gitlab.io/fltk/group__fl__events.html#ga1f6eb61b05d03838978c6a1a8016ce7a) says: _"Sets the widget that will receive FL_KEYBOARD events."_

Whereas [documentation of `Fl_Widget::set_visible_focus()`](https://fltk.gitlab.io/fltk/classFl__Widget.html#ab1fe36207b4a59cae3796b0e6d3b7df8) says: _"Enables **keyboard focus navigation** with this widget."_ and there's also `Fl_Widget::take_focus()` which _"Tries to make this widget be the Fl::focus() widget, ..."_ and correctly checks the `visible_focus` attribute before it calls `Fl::focus()`.

**Comment by Albrecht-S** [comment]
I'm still investigating, but it seems that the offending statement has been introduced when `Fl_Widget::visible_focus()` had been added in FLTK 1.1 (commit 57cef6a4bde017e205e54271e769de0db596d893) but this was IMHO a bug right from the beginning. Maybe this was done because it appeared as a simple solution but it prevents setting the focus to a widget when only "keyboard navigation" should be disabled.

An example is `Fl_Simple_Terminal` which should not receive the keyboard focus by navigation with `TAB` etc. but must still be able to receive the focus by mouse click so the user can copy text to the clipboard. See #365.

**Comment by Albrecht-S** [comment]
@michaelrsweet Mike, can you as the author of commit 57cef6a4bde017e205e54271e769de0db596d893 maybe shed some light on this issue? I know it's almost 20 years ago, but anyway. Maybe it was just "one check too much"?

**Comment by michaelrsweet** [comment]
@Albrecht-S I think the issue was preventing a non-focusable widget getting first focus in a window, but I'm not 100% sure on that. I'm guessing we discussed this on the old mailing list, but those archives are long gone... :/

**Comment by Albrecht-S** [comment]
> I think the issue was preventing a non-focusable widget getting first focus in a window, but I'm not 100% sure on that.

@michaelrsweet Thanks for your quick reply. My take on this issue is that calling `Fl_Widget::clear_visible_focus()` disables giving this widget the focus entirely which was probably not intended. It should only disable taking the focus by using **keyboard navigation** with `TAB` (as documented) but explicitly giving the focus to a widget by calling `Fl::focus(Fl_Widget*)` should still work. `Fl_Text_Display` and derived classes (e.g. `Fl_Text_Editor` and `Fl_Simple_Terminal`) would do this on mouse click events - which does no longer work if `"visible focus"` is disabled.

 I'll look further into it...

**Comment by Albrecht-S** [comment]
This issue should be fixed or (if not possible) documented appropriately before we release FLTK 1.4.0.

**Comment by ManoloFLTK** [comment]
I confess I'm lost about what's the purpose of Fl_Widget::visible_focus(int) .

1) Its name suggests it's related to the widget being drawn with a dotted line when it's focused.
That's indeed the case if you apply ->visible_focus(0) to a Fl_Toggle_Button widget for example :
with this, the dotted line disappears when the button gets redrawn atfer having been pressed.

2) Its documentation mentions something entirely different, that this function modifies the keyboard focus navigation.

3) Its effect mixes both aspects, as seen modifying test/input.cxx with
`b->visible_focus(0);`
added after
` b = new Fl_Toggle_Button(10,y,200,25,"FL_WHEN_CHANGED");`     (line # 120)

- Run first the unmodified input prog, put focus to the top text field, click mouse on "FL_WHEN_CHANGED": the button contains the dotted line, and the top text field has lost focus;
- Run the modified input prog, put focus and click mouse as above : the clicked button has no dotted line but the focus stays in the top text field which receives text that you may type-in then. Clearly, the focus stays unchanged because of the test done inside Fl::focus(Fl_Widget*) being discussed here.

Do we want Fl_Widget::visible_focus(int) to have both these effects? 

Do we need something that prevents focus being transmitted to a widget, knowing that
the widget's handle method returning 0 to the FL_FOCUS event does that already ?

**Comment by Albrecht-S** [comment]
1. The tooltip of the global options in fluid describes the purpose of `OPTION_VISIBLE_FOCUS` better than the actual docs:
```
OPTION_VISIBLE_FOCUS

If visible focus is switched on, FLTK will draw a dotted rectangle inside the
widget that will receive the next keystroke. If switched off, no such indicator
will be drawn and keyboard navigation is disabled.
```

Note the last part _"... **and** keyboard navigation is disabled"_.

The docs should make this clear and I'll try to do this (that's why I added the label "documentation").

2. See 1.

3. The effect is correct as far as it does visually indicate the focus (when the widget has the focus) and that it enables keyboard navigation with TAB, i.e. widgets that have `VISIBLE_FOCUS` disabled are skipped when you press TAB.

> Run the modified input prog, put focus and click mouse as above : the clicked button has no dotted line but the focus stays in the top text field ...

In my version the clicked button **has** the dotted line which is another issue and I found a fix for it. I'm attaching my current patch which is still work in progress.
[issue-366_focus.diff](https://github.com/fltk/fltk/files/7856659/issue-366_focus.diff.txt)

This patch includes the suggested fix in Fl::focus() which is not yet fully tested. I also had to make `unsigned int visible_focus()` `const` which is another fix (this should have been done anyway).

> Do we want Fl_Widget::visible_focus(int) to have both these effects?

Yes, definitely.

> Do we need something that prevents focus being transmitted to a widget, knowing that
the widget's handle method returning 0 to the FL_FOCUS event does that already ?

Yes. The difference is like having to derive a class or to use a callback, in this case you'd have to derive a class and implement the handle() method to reject `FL_FOCUS` events. It's much simpler from a programmer's point of view to call `Fl_Widget::visible_focus(0)` in cases when this is needed or wanted.

The docs should also clearly state that you can still set the focus to the widget if you disable `VISIBLE_FOCUS` by calling for instance `Fl::focus(this)` or similar. This is not possible in current FLTK but works with my patch.

Yet another test case: set `visible_focus(0)` on one of the input fields. You'll notice that this field will be skipped by tab navigation. However, you can't click into the field to set the focus and start typing in the unmodified FLTK lib. My patch fixes this as well. The same applies to `Fl_Text_Display` and its derived class `Fl_Simple_Terminal` which was the root cause of issue #365 and led me to open this one (#366).

I'm still investigating but test results with my posted patch would be very much appreciated. Does this patch have any unwanted side effects?

**Comment by erco77** [comment]
On 1/12/22 6:42 AM, ManoloFLTK wrote:
>
> I confess I'm lost about what's the purpose of Fl_Widget::visible_focus(int) .
>

     As I recall, with respect to button widgets, it was the way to disable the focus box
     from being drawn on e.g. button like widgets where the focus box is inappropriate
     to see due to the needs of the application, such as buttons that are flat or invisible,
     and not intended to be seen as a button (e.g. invisible buttons positioned over images
     or other graphics, or buttons that had to be clicked to be activated, and never pushed
     by keyboard activity, etc)

     IIRC disabling this had the effect of disabling the widget from taking focus,
     perhaps to prevent accidental triggering by a stray keyhit because there's
     no visual feedback to the user that the widget has focus.

     I'm only familiar with this feature in the context of buttons, not input widgets.

**Comment by Albrecht-S** [comment]
@erco77 wrote:
>  I'm only familiar with this feature in the context of buttons, not input widgets.

Yes, I think this is the way it's implemented: only for non-input widgets like buttons etc.. Hence I withdraw my "request" (see title _"Fl::focus(Fl_Widget*) should be independent of Fl_Widget::visible_focus()"_) because changing this would mean to redefine the general behavior and my investigation of existing code shows that this would very likely introduce unexpected behavior changes.

Hence the only remaining aspect of this issue is to update the documentation and maybe the tooltip in fluid. I'll change the title accordingly.

**Comment by leenzhu** [comment]
I just want to disable drawing the dotted rectangle inside some widgets for better appearence  (eg. browser item, tree item) while keeping keyboard navigation. 

But as the document says:

> If visible focus is switched on, FLTK will draw a dotted rectangle inside the
widget that will receive the next keystroke. If switched off, no such indicator
will be drawn and **keyboard navigation is disabled**.

Both Fl::focus(Fl_Widget*) and Fl_Widget::visible_focus() will disable the keybord navigation. 

After investigating the codes of fltk-1.4.x-20230512-7a50c823, I havn't found any way just preventing draw dotted rectangle.

In my option, it would be better if there were two methods:

1. Fl::visible_foucs() for drawing dotted rectagle or not.
2. Fl::enable_foucs() for keybord navigation or not. Ofcause, if dissabled, dotted rectangel neither drawn.

**Comment by Albrecht-S** [comment]
@leenzhu Thanks for your comment and suggestion. I see your issue and agree that we might want to add another method that just inhibits drawing the focus rectangle (likely only per widget) but does not change the keyboard focus behavior. This should be investigated further.

You wrote:
> Both Fl::focus(Fl_Widget*) and Fl_Widget::visible_focus() will disable the keybord navigation.

This statement doesn't seem to be correct (typo or copy/paste error?):

- [Fl::focus(Fl_Widget*)](https://www.fltk.org/doc-1.4/group__fl__events.html#ga1f6eb61b05d03838978c6a1a8016ce7a) sets the widget that (currently) has the keyboard focus (gets keyboard events). This does not change keyboard navigation behavior.
- [Fl_Widget::visible_focus(int)](https://www.fltk.org/doc-1.4/classFl__Widget.html#a6a2a0769c73831e6ef0076d9efa98d72) enables or disables the keyboard focus behavior **per widget**.

That said,
- [Fl::visible_focus(int)](https://www.fltk.org/doc-1.4/classFl.html#ade5fd097d36257dce2b7aff1165e037f) enables or disables the keyboard focus behavior **globally**.

Note that we can't change the behavior of `Fl::visible_foucs()` because of backwards compatibility, hence this method will always do both, enable/disable keyboard navigation **and** enable/disable drawing the focus rectangle.

However, we can add new methods (either global: `Fl::any_name()` or per widget: `Fl_Widget::any_name()`) to enable and disable drawing the focus rectangle (replace `any_name` with a new method name).

**Comment by leenzhu** [comment]
Woops, sorry for the typo. You are right, I actually mean `Fl::visible_focus()` rather than `Fl::foucs()`.

I agree with you that visible_foucs() should keep the behavior the same as before for backwards compatibility. I use the term `visible_` for `vison` and `enable_` for `action`, the two names are used for the purpose of distinguishing between `vison` and `action`. I should have explained in more detail.

**Comment by Albrecht-S** [comment]
@MatthiasWM Is there any progress or do you have any comments on this issue?

After reading this entire issue and taking a look at my proposed patch I believe that this patch is worth being considered. At least some (benign) parts like making `Fl_Widget::visible_focus()` **_const_**.

The other point that calling `Fl::focus(Fl_Widget *)` refuses to set the focus if `visible_focus()` is turned off is IMHO really questionable.

Matthias, can you please take look at my patch? I know it was not completely tested but you may see something or agree to commit it. I would at least like to get a second opinion before I (or you) want to commit anything like this shortly before the release.

With or without the patch (bug fix) we should try to document what these functions really do...

**Comment by MatthiasWM** [comment]
tl;dr: no further changes now, in 1.4, rename to `enable_focus()` or something deprecate `visible_focus()`, make `draw_focus()` virtual, and finally add a `draw_focus_f` callback to all box types.

Don't know what to say here. I find the idea of invisible_focus quite useless, and I have a hard time constructing situations where this makes sense and isn't covered by returning 0 from FL_FOCUS events. But it's here and it needs to be resolved. @leenzhu is absolutely correct that we use one flag for two quite different purposes. 

The functional aspect of not taking focus is mostly covered by deactivating a widget. I can somewhat see how it may make sense that a widget can not get keyboard focus while still active for mouse use, but then it should also not be possible to force focus with `Fl::focus(Fl_Widget*)` either.

The visual aspect is different (an in that respect, `visible_focus()` is a very bad name choice). The focus rect looks bad on many box types, but I have that on the list of fixes (STR 2145) by adding a draw_focus callback function to box types. This would also allow better visualization than the dotted rectangle. If that still does not help to achieve a certain design, the user can still override `::draw()`. Maybe `::draw_focus()` could be virtual to give more options here.

**Comment by Albrecht-S** [comment]
> I have a hard time constructing situations where this makes sense and isn't covered by returning 0 from FL_FOCUS events

The difference is that users can set `visible_focus(0)` in their GUI w/o deriving a new class. AFAICT the original purpose was to be used on buttons and similar widgets (not on input widgets which usually don't show a focus rect). Imagine a GUI where you have several buttons intermixed with input fields. If you set `visible_focus(0)` on all buttons you can walk from input field to input field by pressing TAB w/o stopping at each button. Buttons would have to be clicked with the mouse. I can even imagine a scenario where each input field has several buttons left or right of it (maybe check or toggle buttons) etc..

As I wrote above this feature was added in FLTK 1.1 in commit 57cef6a4bde017e205e54271e769de0db596d893 but the commit log doesn't say **why** it was added. I assume it was done on user request.

Other than that I agree that we shouldn't make functional changes in 1.4, we should hold this back for 1.5.

However, wouldn't it make sense to make the getter `unsigned int visible_focus()` at least `const` as in the patch below? I don't think it would do any harm and it would be **_correct_** because this method doesn't change the widget.

Proposed patch:
```diff
--- a/FL/Fl_Widget.H
+++ b/FL/Fl_Widget.H
@@ -845,7 +845,7 @@ public:
       \retval 0 if this widget has no visible focus.
       \see visible_focus(int), set_visible_focus(), clear_visible_focus()
    */
-  unsigned int visible_focus() { return flags_ & VISIBLE_FOCUS; }
+  unsigned int visible_focus() const { return flags_ & VISIBLE_FOCUS; }
 
   /** The default callback for all widgets that don't set a callback.
 
```

@MatthiasWM What do you think?

**Comment by Albrecht-S** [comment]
Note: I removed this issue from "Release FLTK 1.4.0" milestone because there's no action required now. Leaving it open for later.

**Comment by MatthiasWM** [comment]
Is this still active in any way?

This issue will expire on 23 Nov 2025

**Comment by Albrecht-S** [comment]
The main issue is IMHO that `visible_focus` was implemented with two things in mind:

1. to draw or not to draw the focus box when the widget has the focus
2. to skip the widget when moving the focus from one widget to the next using the keyboard (by pressing tab or shift-tab)

The existing code seems to have some undesirable side effects (IIRC) which adds to the confusion about what this method should do and what not.

My intention was to **clarify** this - at least in the docs - but currently I don't think that's doable because **it is not clear** what the intention is (was) and what it actually does. :-(

Maybe we can revisit this issue later and define a proper API etc. etc..

Anyway, I'm closing this issue now.
